app/grandchallenge/reader_studies/tasks.py

```diff
@@ -37,6 +46,43 @@ def add_scores(*, instance_pk, pk_set):
             add_score(instance, ground_truth.answer)
 
 
+@shared_task(**settings.CELERY_TASK_DECORATOR_KWARGS["acks-late-micro-short"])
+def add_image_to_display_set(
+    *, upload_session_pk, display_set_pk, interface_pk
+):
+    display_set = DisplaySet.objects.get(pk=display_set_pk)
+    image = Image.objects.get(origin_id=upload_session_pk)
+    interface = ComponentInterface.objects.get(pk=interface_pk)
+    with transaction.atomic():
+        display_set.values.remove(
+            *display_set.values.filter(interface=interface)
+        )
+        civ, _ = ComponentInterfaceValue.objects.get_or_create(
+            interface=interface, image=image
+        )
+        display_set.values.add(civ)
+
+
+@shared_task(**settings.CELERY_TASK_DECORATOR_KWARGS["acks-late-2xlarge"])
+def create_display_sets_for_upload_session(
+    *, upload_session_pk, reader_study_pk, interface_pk
+):
+    images = Image.objects.filter(origin_id=upload_session_pk)
+    reader_study = ReaderStudy.objects.get(pk=reader_study_pk)
+    interface = ComponentInterface.objects.get(pk=interface_pk)
+    with transaction.atomic():
+        for image in images:
+            civ, _ = ComponentInterfaceValue.objects.get_or_create(
+                interface=interface, image=image
+            )
+            if DisplaySet.objects.filter(
+                reader_study=reader_study, values=civ
+            ).exists():
+                continue
```

Comment:
I thought that we wanted to allow re-use of CIVs in a study?

Reply:
We do, but in this case the only way a display set for this civ can already exist is if this task gets called twice. This task assigns a newly created upload to a civ, so this is the proper behavior here imo.